File: location.ts
const readline = require('readline');

interface Car {
    id: number;
    name: string;
    reserved: boolean;
    days: number;
    price: number;
}

const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout,
});

const cars: Car[] = [
    { id: 1, name: 'Peugeot 208', reserved: false, days: 0, price: 40 },
    { id: 2, name: 'Toyota Corolla', reserved: false, days: 0, price: 45 },
    { id: 3, name: 'Renault Clio', reserved: false, days: 0, price: 35 },
    { id: 4, name: 'Ford Fiesta', reserved: false, days: 0, price: 50 },
    { id: 5, name: 'Volkswagen Golf', reserved: false, days: 0, price: 55 },
];

function ask(question: string): Promise<string> {
    return new Promise((resolve) => rl.question(question, resolve));
}

function displayCars(): void {
    console.log('\nListe des voitures :');
    cars.forEach((car) => {
        const status = car.reserved
            ? `Réservée (${car.days} jrs)`
            : 'Disponible';
        console.log(`[${car.id}] ${car.name} - ${status}`);
    });
}

async function reserveCar(): Promise<void> {
    const input = await ask("\nEntrez l'ID ou le nom de la voiture : ");
    const car = cars.find(
        (c) =>
            c.id.toString() === input ||
            c.name.toLowerCase() === input.toLowerCase()
    );

    if (!car) {
        console.log('\nxxx Voiture introuvable xxx');
        return;
    }

    if (car.reserved) {
        console.log('\nxxx Cette voiture est déjà réservée xxx');
        return;
    }

    const daysStr = await ask('Nombre de jours de location (1-30) : ');
    const days = parseInt(daysStr);

    if (isNaN(days) || days < 1 || days > 30) {
        console.log('\nxxx Nombre de jours invalide xxx');
        return;
    }

    const confirm = await ask(
        `Confirmer la réservation de ${car.name} pour ${days} jours à ${car.price}€/jour ? (o/n) : `
    );

    if (confirm.toLowerCase() !== 'o') {
        console.log('\nxxx Réservation annulée xxx');
        return;
    }

    car.reserved = true;
    car.days = days;
    const total = car.price * days;

    console.log(`\n------------------------\n\n✅ RESERVATION CONFIRMEE !
Voiture : ${car.name}
Durée : ${days} jours
Total : ${total.toFixed(2)} €
\n------------------------`);
}

async function main(): Promise<void> {
    console.log("\nBienvenue dans l'application de location de voitures !");

    while (true) {
        console.log('\n--- MENU PRINCIPALE ---');
        displayCars();

        console.log('\nOptions :');
        console.log('1 - RESERVER UNE VOITURE');
        console.log('2 - QUITTER');

        const choice = await ask('Choix : ');

        if (choice === '1') {
            await reserveCar();
        } else if (choice === '2') {
            console.log('\nÀ bientôt !\n');
            break;
        } else {
            console.log('xxx Choix invalide !!! Veuillez réessayer.');
        }
    }

    rl.close();
}

main();

File: location.py
# --- app location with python ---

from dataclasses import dataclass
from typing import List
from colorama import Fore, Style, init

init(autoreset=True)

@dataclass
class Car:
    id: int
    name: str
    reserved: bool
    days: int
    price: float

cars: List[Car] = [
    Car(1, "Peugeot 208", False, 0, 40),
    Car(2, "Toyota Corolla", False, 0, 45),
    Car(3, "Renault Clio", False, 0, 35),
    Car(4, "Ford Fiesta", False, 0, 50),
    Car(5, "Volkswagen Golf", False, 0, 55),
]

def display_welcome():
    print(Fore.CYAN + "\nBienvenue dans l'application de location de voitures !")

def display_menu():
    print("\n--- MENU PRINCIPAL ---")
    print("1 - RESERVER UNE VOITURE")
    print("2 - QUITTER")

def display_cars():
    print("\nListe des voitures disponibles :\n")
    for car in cars:
        status = (
            Fore.RED + f"Réservée ({car.days} jrs)"
            if car.reserved else Fore.GREEN + "Disponible"
        )
        print(f"[{car.id}] {car.name:<20} - {status}")

def find_car_by_input(input_str: str):
    return next(
        (c for c in cars if str(c.id) == input_str or c.name.lower() == input_str.lower()),
        None
    )

def reserve_car():
    input_str = input("\n Entrez l'ID ou le nom de la voiture : ").strip()
    car = find_car_by_input(input_str)

    if not car:
        print(Fore.YELLOW + "xxx Voiture introuvable xxx")
        return

    if car.reserved:
        print(Fore.YELLOW + "xxx Cette voiture est déjà réservée xxx")
        return

    try:
        days = int(input(" Nombre de jours de location (1-30) : ").strip())
        if not 1 <= days <= 30:
            print(Fore.YELLOW + "xxx Le nombre de jours doit être entre 1 et 30 xxx")
            return
    except ValueError:
        print(Fore.YELLOW + "xxx Entrée invalide xxx")
        return

    confirm = input(f" ? Confirmer la réservation de {car.name} pour {days} jours à {car.price}€/jour ? (o/n) : ").lower().strip()
    if confirm != 'o':
        print(Fore.YELLOW + " !!! Réservation annulée !!!")
        return

    car.reserved = True
    car.days = days
    total = car.price * days

    print(Fore.GREEN + "\n✅ RESERVATION CONFIRMÉE")
    print("-" * 40)
    print(f"Voiture : {car.name}")
    print(f"Durée   : {days} jours")
    print(f"Total   : {total:.2f} €")
    print("-" * 40)

def main():
    display_welcome()

    while True:
        display_cars()
        display_menu()

        choice = input("\n🡒 Votre choix : ").strip()

        if choice == "1":
            reserve_car()
        elif choice == "2":
            print(Fore.CYAN + "\n Merci et à bientôt ! \n")
            break
        else:
            print(Fore.YELLOW + "!!! Choix invalide. Réessayez !!!")

if __name__ == "__main__":
    main()
